src/PowerPing/Ping.cs

```diff
@@ -410,24 +417,39 @@ private PingResults SendICMP(PingAttributes attrs)
                         if (rnd.Next(20) == 1) { throw new SocketException(); }
                     }
 
-                    // Wait for response
-                    byte[] buffer = new byte[attrs.RecieveBufferSize]; // Ipv4Header.length + IcmpHeader.length + attrs.recievebuffersize
-                    bytesRead = sock.ReceiveFrom(buffer, ref ep);
-                    responseTimer.Stop();
-
-                    // Store reply packet
-                    ICMP response = new ICMP(buffer, bytesRead);
+                    ICMP response;
+                    TimeSpan replyTime;
+                    do {
+                        // Wait for response
+                        bytesRead = sock.ReceiveFrom(receiveBuffer, ref ep);
+                        replyTime = new TimeSpan(Helper.StopwatchToTimeSpanTicks(Stopwatch.GetTimestamp() - requestTimestamp));
+
+                        // Store reply packet
+                        response = new ICMP(receiveBuffer, bytesRead);
+
+                        // Ignore unexpected echo responses
+                        if (packet.type == 8 && response.type == 0) {
```

Comment:
I think this check makes sense when we are sending standard request packets, but when we are sending custom packet types we might not recieve type 0 or 8 packets. Maybe we should disable this check when a custom packet type is being used.

Reply:
That was the goal of the logic there; the validation only happens if packet.type (request type, which comes from attrs.Type) is 8, AND response.type is 0.

Reply:
Oh I see, my bad, must have misread it. That works great then.